src/irclog2html/irclogsearch.py

```diff
@@ -236,39 +243,89 @@ def print_search_results(query, where=None):
     if prev_result:
         formatter.print_suffix()
     if date:
-        print("  </li>")
-        print("</ul>")
+        print("  </li>", file=stream)
+        print("</ul>", file=stream)
     total_time = time.time() - started
     print("<p>%d matches in %d log files with %d lines (%.1f seconds).</p>" % (
-                stats.matches, stats.files, stats.lines, total_time))
-    print(FOOTER)
+                stats.matches, stats.files, stats.lines, total_time),
+          file=stream)
+    print(FOOTER, file=stream)
+
+    return formatter # destroying it closes the stream, breaking the result
 
 
 def rewrap_stdout():
-    # XXX: shouldn't replace sys.stdout, should pass the output stream to
-    # the functions that want to print stuff
     if hasattr(sys.stdout, 'buffer'):
         stream = sys.stdout.buffer # Python 3
     else:
         stream = StdoutWrapper(sys.stdout) # Python 2
-    sys.stdout = io.TextIOWrapper(stream, 'ascii',
-                                  errors='xmlcharrefreplace',
-                                  line_buffering=True)
+    return io.TextIOWrapper(stream, 'ascii',
+                            errors='xmlcharrefreplace',
+                            line_buffering=True)
+
+
+def search_page(stream, form, where, logfile_pattern):
+    if "q" not in form:
+        print_search_form(stream)
+    else:
+        search_text = form["q"].value
+        if isinstance(search_text, bytes):
+            search_text = search_text.decode('UTF-8')
+        return print_search_results(search_text, stream=stream, where=where,
+                                    logfile_pattern=logfile_pattern)
+
+
+def get_path(environ):
+    path = environ.get('PATH_INFO', '/')
+    path = path.split('/')[-1]  # is this enough to protect from FS traversal?
+    return path if path else 'index.html'
+
+
+def wsgi(environ, start_response):
+    """WSGI application"""
+    logfile_path = environ.get('IRCLOG_LOCATION')
+    logfile_pattern = environ.get('IRCLOG_GLOB')
+    form = cgi.FieldStorage(fp=environ['wsgi.input'], environ=environ)
+    stream = io.TextIOWrapper(io.BytesIO(), 'ascii',
+                              errors='xmlcharrefreplace',
+                              line_buffering=True)
+
+    status = str("200 Ok")
+    content_type = str("text/html; charset=UTF-8")
```

Comment:
Could be text/plain, if you try to access the .log file directly.  Charset is tricky: log files might be in xchat's hybrid UTF-8/Latin-1 mix.